src/features/Cameras/scripts/lib/canvas/virtualButtons.ts

```diff
@@ -4,77 +4,130 @@ import {
+
+  height_rate_sum: Array(),
+  width_rate_sum: Array(),
+};
+
+export function buttons_draw_grid() {
+  // Canvas要素を取得
+  const canvas = getCanvasElement();
+  const context = getCanvasContext(canvas);
+
+  virtualButtons.height_rate_sum = [0];
+  virtualButtons.width_rate_sum = [0];
+  let cnt = 0;
+  for (let i = 0; i < virtualButtons.grid_width; i++) {
+    cnt += virtualButtons.width_rate[i];
+    virtualButtons.width_rate_sum.push(cnt);
+  }
+  cnt = 0;
+  for (let i = 0; i < virtualButtons.grid_height; i++) {
+    cnt += virtualButtons.height_rate[i];
+    virtualButtons.height_rate_sum.push(cnt);
+  }
+
+  for (let i = 0; i < virtualButtons.grid_width; i++) {
+    let x = canvas.width * virtualButtons.width_rate_sum[i];
+
+    context.beginPath();
+    context.moveTo(x, 0); //始点
+    context.lineTo(x, canvas.height); //終点
+    context.closePath();
+    context.stroke();
+  }
+  for (let i = 0; i < virtualButtons.grid_height; i++) {
+    let y = canvas.height * virtualButtons.height_rate_sum[i];
+
+    context.beginPath();
+    context.moveTo(0, y); //始点
+    context.lineTo(canvas.width, y); //終点
+    context.closePath();
+    context.stroke();
+  }
+}
+
+export function buttons_draw(poses: Pose[]) {
+  // Canvas要素を取得
+  const canvas = getCanvasElement();
+  const context = getCanvasContext(canvas);
+
+  let keypoints: Keypoint[] = poses.length != 0 ? poses[0].keypoints : Array();
+
+  let cnt = 0;
+  for (let i = 0; i < virtualButtons.grid_height; i++) {
+    cnt += virtualButtons.height_rate[i];
+    virtualButtons.height_rate_sum.push(cnt);
+  }
+  cnt = 0;
+  for (let i = 0; i < virtualButtons.grid_width; i++) {
+    cnt += virtualButtons.width_rate[i];
+    virtualButtons.width_rate_sum.push(cnt);
+  }
+
+  // 仮想ボタンの描画
+  for (let i = 0; i < virtualButtons.grid_height; i++) {
+    for (let j = 0; j < virtualButtons.grid_width; j++) {
+      if (!virtualButtons.buttons_visibility[i][j]) continue;
+
+      let x1, y1, x2, y2: number;
+      y1 = canvas.height * virtualButtons.height_rate_sum[i];
+      x1 = canvas.width * virtualButtons.width_rate_sum[j];
+      y2 = canvas.height * virtualButtons.height_rate_sum[i + 1];
+      x2 = canvas.width * virtualButtons.width_rate_sum[j + 1];
+
+      context.fillStyle = virtualButtons.colors[0];
+
+      for (let target of virtualButtons.valid_keypoints) {
+        if (contains(i, j, keypoints, target)) {
+          context.fillStyle = virtualButtons.colors[1];
         }
       }
 
-      context.fillRect(x, y, l1, l2);
+      context.fillRect(x1, y1, x2 - x1, y2 - y1);
     }
-  },
-};
+  }
+}
+
+function findKeypointByName(
+  keypoints: Keypoint[],
+  targetName: String,
+): Keypoint | undefined {
+  return keypoints.find((keypoint) => keypoint.name === targetName);
+}
+
+function contains(h: number, w: number, keypoints: Keypoint[], target: String) {
+  let pnt: Keypoint = findKeypointByName(keypoints, target) as Keypoint;
```

Comment:
`as Keypoint` と型宣言はいらないと思う
`as Keypoint | undefined` という書き方なら理解できるが、それなら書く必要はないし

```suggestion
  let pnt = findKeypointByName(keypoints, target);
```

Reply:
@mznms さんに直してもらうまで、`pnt===undefined`の時の場合分けをしていなかったので、無理やりそう書いていました。
修正します！

Reply:
将成も一枚噛んでるのね
了解です